### ISSUE TITLE
[12.0] getTemporaryTexture returning dirty stencil texture, causing ugly artifacts.

### ISSUE THREAD

**Issue by Klonan** [love2d/love#1843]
When rendering with stencils, if I did many in a single frame, it would cause artifacts and stencil failures:
![love_uN3FTdRrVJ](https://user-images.githubusercontent.com/11986037/190876075-99290e3f-0f3c-47a3-90b2-da49b0d9051f.png)![love_S282CIdUbH](https://user-images.githubusercontent.com/11986037/190876076-c35ec84f-8400-431c-8491-5e8e538b9481.png)
![love_9Up7tTj7cc](https://user-images.githubusercontent.com/11986037/190876078-3230c3ea-81aa-4d45-9d17-4d6124f18aa7.png)

If I do the stencils slowly, its fine, fastly, its glitched:

https://user-images.githubusercontent.com/11986037/190876129-3b343310-10af-45b8-9fa9-78cf38ba989e.mp4

Which pointed to this magic number about frames: 
![Code_NBg6vrGA1s](https://user-images.githubusercontent.com/11986037/190876142-84c6820d-c166-4f42-a374-06463472dfaf.png)

And from there I traced it down to the `getTemporaryTexture`, which is returning a dirty texture:
![Code_K0XG7DDqyk](https://user-images.githubusercontent.com/11986037/190876096-b8ae20b0-3834-48c9-9641-6c8a89171f15.png)

**Comment by slime73** [comment]
The stencil buffer is cleared when you call `love.graphics.clear`, and also when you call `love.graphics.stencil` (as long as you don't set `keepvalues` to true).

It sounds like you might be setting keepvalues to true in every stencil() call in a frame, without clearing the stencil buffer via clear()? In that situation since you're intentionally avoiding resetting it to a specific state, it may have undefined contents.

**Comment by Klonan** [comment]
In 12.0 the love.graphics.stencil was replaced with the more generic `setStencilMode()`,

My code is something like this: (`Graphics` is just a local variable of `love.graphics`)
```lua
  self.shapeCanvas = Graphics.newCanvas(w + canvasTolerance, h + canvasTolerance)
  Graphics.setCanvas({self.shapeCanvas, stencil = true})
  Graphics.setStencilMode("increment", "always")
  Graphics.translate(canvasOffset - x1, canvasOffset - y1)
  for k, shape in pairs(self.shapes) do
    if shape.type == "polygon" then
      Graphics.polygon("fill", shape.points)
    elseif shape.type == "circle" then
      local x, y = shape.x or 0, shape.y or 0
      Graphics.circle("fill", x, y, shape.radius)
    end
  end
  Graphics.setStencilMode("replace", "greater", 0)
```

I fixed the issue by calling `Graphics.clear()` after the first `setCanvas`, but I would expect since I am passing a new/fresh canvas to use as the stencil each time, it would not persist any previous values from unrelated canvases/draw calls

**Comment by slime73** [comment]
> I fixed the issue by calling Graphics.clear() after the first setCanvas, but I would expect since I am passing a new/fresh canvas to use as the stencil each time, it would not persist any previous values from unrelated canvases/draw calls

The active color canvas isn't related to / doesn't own the stencil buffer. If you want to have a stencil canvas you pass around, you can do that (by calling newCanvas with a stencil format like `"stencil8"`, and then setting `depthstencil = stencilcanvas` in setCanvas).

**Comment by Klonan** [comment]
Okay, I think I just was not 100% on how the stencil and canvas functions work in 12.0, if I get it now, the setCanvas by a by-product, also creates a temporary stencil buffer, which based on the width/height, is reused for other stencil usage

Since this is a transparent effect, I would expect the stencil to be cleared automatically in the setCanvas call. It is not always possible to call the `graphics.clear()' after the setCanvas, as you may want to mask out the same canvas and paint to it in multiple steps. A sub-optimal solution then it to make the user made stencilCanvas, and switch to it, clear it, then switch back to the colorCanvas. 

The goal I suppose of the auto generated stencil canvas is to make things easier and 'just work' without added steps or gotchas, so not auto-clearing is a hazard to that

**Comment by slime73** [comment]
> It is not always possible to call the `graphics.clear()' after the setCanvas, as you may want to mask out the same canvas and paint to it in multiple steps.

There are love.graphics.clear variants which let you only clear the depth/stencil buffers if you choose. But I agree that overall it's not very intuitive right now when auto-generated stencil buffers are used.

**Comment by slime73** [comment]
> It might make more sense for love to automatically clear the stencil buffer inside setCanvas (when it's not a manually created / user-managed one), although if people call love.graphics.clear right after that it'd result in a bit worse performance because of the redundant clear call in that situation.

This is implemented now in [94d05e3](https://github.com/love2d/love/commit/94d05e319b8b672683e735ef3126950abdd0d607).
